[spmd] Add list handling to data parallel and add foreach tests

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #99374
* #99489
* #99436
* __->__ #99373
* #99435
* #99062
* #98452

This PR adds list handling logic to the new DataParallel expansion and
add foreach optimizer tests, currently current testing sgd optimizers
in foreach mode, for both replicate and fully shard

Next step:

Add fused optim tests

wanchaol committed:
[spmd] Add list handling to data parallel and add foreach tests

This PR adds list handling logic to the new DataParallel expansion and
add foreach optimizer tests, currently current testing sgd optimizers
in foreach mode, for both replicate and fully shard

Next step:

Add fused optim tests

[ghstack-poisoned]

wanchaol committed:
Update on "[spmd] Add list handling to data parallel and add foreach tests"


This PR adds list handling logic to the new DataParallel expansion and
add foreach optimizer tests, currently current testing sgd optimizers
in foreach mode, for both replicate and fully shard

Next step:

Add fused optim tests

[ghstack-poisoned]

wanchaol committed:
Update on "[spmd] Add list handling to data parallel and add foreach tests"


This PR adds list handling logic to the new DataParallel expansion and
add foreach optimizer tests, currently current testing sgd optimizers
in foreach mode, for both replicate and fully shard

Next step:

Add fused optim tests

[ghstack-poisoned]

mrshenli commented on torch/distributed/_spmd/data_parallel.py:
```diff
@@ -261,13 +262,14 @@ def build_data_parallel_strategies(
         aten.reshape.default,
         aten.detach.default,
         aten.clone.default,
+        torch.ops._spmd.tag_grad.default,
```
nit: aten ops above are not ordered either :)

mrshenli commented on torch/distributed/_spmd/data_parallel.py:
```diff
@@ -367,7 +368,10 @@ def build_data_parallel_strategies(
-                if arg_node_type == NodeType.PARAM:
+                if node.target == operator.getitem:
+                    # for getitem call, just forward the strategy from the input
+                    dp_strategy_map[node] = dp_strategy_map[node.args[0]]
```
hmm, input here is a list of nodes and output might be a list or tensor. So do we mark the foreach output as a PARAM or NON_TENSOR?

mrshenli replied:
Also, does this mean individual tensors from foreach output does not have a dp strategy associated with it?

mrshenli replied:
nvm, I saw the PRs on top is trying to address this problem

wanchaol committed:
Update on "[spmd] Add list handling to data parallel and add foreach tests"


This PR adds list handling logic to the new DataParallel expansion and
add foreach optimizer tests, currently current testing sgd optimizers
in foreach mode, for both replicate and fully shard

Next step:

Add fused optim tests

[ghstack-poisoned]

wanchaol committed:
Update on "[spmd] Add list handling to data parallel and add foreach tests"


This PR adds list handling logic to the new DataParallel expansion and
add foreach optimizer tests, currently current testing sgd optimizers
in foreach mode, for both replicate and fully shard

Next step:

Add fused optim tests

[ghstack-poisoned]

wanchaol committed:
Update on "[spmd] Add list handling to data parallel and add foreach tests"


This PR adds list handling logic to the new DataParallel expansion and
add foreach optimizer tests, currently current testing sgd optimizers
in foreach mode, for both replicate and fully shard

Next step:

Add fused optim tests

[ghstack-poisoned]

wanchaol committed:
Update on "[spmd] Add list handling to data parallel and add foreach tests"


This PR adds list handling logic to the new DataParallel expansion and
add foreach optimizer tests, currently current testing sgd optimizers
in foreach mode, for both replicate and fully shard

Next step:

Add fused optim tests

[ghstack-poisoned]

wanchaol committed:
Update on "[spmd] Add list handling to data parallel and add foreach tests"


This PR adds list handling logic to the new DataParallel expansion and
add foreach optimizer tests, currently current testing sgd optimizers
in foreach mode, for both replicate and fully shard

Next step:

Add fused optim tests

[ghstack-poisoned]

wanchaol committed:
Update on "[spmd] Add list handling to data parallel and add foreach tests"


This PR adds list handling logic to the new DataParallel expansion and
add foreach optimizer tests, currently current testing sgd optimizers
in foreach mode, for both replicate and fully shard

Next step:

Add fused optim tests

[ghstack-poisoned]

wanchaol committed:
Update on "[spmd] Add list handling to data parallel and add foreach tests"


This PR adds list handling logic to the new DataParallel expansion and
add foreach optimizer tests, currently current testing sgd optimizers
in foreach mode, for both replicate and fully shard

Next step:

Add fused optim tests

[ghstack-poisoned]

wanchaol committed:
Update on "[spmd] Add list handling to data parallel and add foreach tests"


This PR adds list handling logic to the new DataParallel expansion and
add foreach optimizer tests, currently current testing sgd optimizers
in foreach mode, for both replicate and fully shard

Next step:

Add fused optim tests

[ghstack-poisoned]

wanchaol committed:
Update on "[spmd] Add list handling to data parallel and add foreach tests"


This PR adds list handling logic to the new DataParallel expansion and
add foreach optimizer tests, currently current testing sgd optimizers
in foreach mode, for both replicate and fully shard

Next step:

Add fused optim tests

[ghstack-poisoned]

wanchaol committed:
Update on "[spmd] Add list handling to data parallel and add foreach tests"


This PR adds list handling logic to the new DataParallel expansion and
add foreach optimizer tests, currently current testing sgd optimizers
in foreach mode, for both replicate and fully shard

Next step:

Add fused optim tests

[ghstack-poisoned]

wanchaol committed:
Update on "[spmd] Add list handling to data parallel and add foreach tests"


This PR adds list handling logic to the new DataParallel expansion and
add foreach optimizer tests, currently current testing sgd optimizers
in foreach mode, for both replicate and fully shard

Next step:

Add fused optim tests

[ghstack-poisoned]

wanchaol committed:
Update on "[spmd] Add list handling to data parallel and add foreach tests"


This PR adds list handling logic to the new DataParallel expansion and
add foreach optimizer tests, currently current testing sgd optimizers
in foreach mode, for both replicate and fully shard

Next step:

Add fused optim tests

[ghstack-poisoned]

wanchaol commented:
@pytorchbot merge
